topdown/json_bench_test.go

```diff
@@ -7,78 +7,503 @@ package topdown
+			}
+			patchList := ast.NewArray(plArrayObj...)
+
+			b.ResetTimer()
+			b.Run(testName, func(b *testing.B) {
+				runJSONPatchBenchmarkTest(ctx, b, source, patchList)
+			})
+		}
+	}
+
+	// Set case
+	for _, n := range sizes {
+		source := genTestSet(n)
+		for _, m := range sizes {
+			testName := fmt.Sprintf("set-%d-%d", n, m)
+			// Build dataset right before use:
+			plSet := make([]*ast.Term, 0, m*2)
+			// add ops
 			for i := 0; i < m; i++ {
-				idx := strconv.FormatInt(int64(i), 10)
-				paths = append(paths, ast.StringTerm("/"+idx+"/"+idx+"/"+idx))
+				plSet = append(plSet, genTestJSONPatchObject("add", ast.ArrayTerm(ast.IntNumberTerm(i+n)), nil, ast.IntNumberTerm(i+n)))
+			}
+			// remove ops
+			for i := m - 1; i >= 0; i-- {
+				plSet = append(plSet, genTestJSONPatchObject("remove", ast.ArrayTerm(ast.IntNumberTerm(i+n)), nil, nil))
 			}
+			patchList := ast.NewArray(plSet...)
 
 			b.ResetTimer()
 			b.Run(testName, func(b *testing.B) {
-				runJSONFilterBenchmarkTest(ctx, b, source, paths)
+				runJSONPatchBenchmarkTest(ctx, b, source, patchList)
 			})
 		}
 	}
 }
 
-func runJSONFilterBenchmarkTest(ctx context.Context, b *testing.B, source ast.Value, paths []*ast.Term) {
+func genTestJSONPatchObject(op string, path, from, value *ast.Term) *ast.Term {
+	patchObj := ast.NewObject(
+		[2]*ast.Term{ast.StringTerm("op"), ast.StringTerm(op)},
+		[2]*ast.Term{ast.StringTerm("path"), path},
+	)
+	if from != nil {
+		patchObj.Insert(ast.StringTerm("from"), from)
+	}
+	if value != nil {
+		patchObj.Insert(ast.StringTerm("value"), value)
+	}
+	return ast.NewTerm(patchObj)
+}
+
+func genTestObject(width int) ast.Value {
+	out := ast.NewObject()
+	for i := 0; i < width; i++ {
+		out.Insert(ast.IntNumberTerm(i), ast.IntNumberTerm(i))
+	}
+	return out
+}
+
+func genTestArray(width int) ast.Value {
+	out := ast.NewArray()
+	for i := 0; i < width; i++ {
+		out = out.Append(ast.IntNumberTerm(i))
+	}
+	return out
+}
+
+func genTestSet(width int) ast.Value {
+	out := ast.NewSet()
+	for i := 0; i < width; i++ {
+		out.Add(ast.IntNumberTerm(i))
+	}
+	return out
+}
+
+// For the purposes of addressing the original Github issue (#4409), a
+// fairly shallow object with many keys ought to do the trick.
+func gen3LayerObject(l1Keys, l2Keys, l3Keys int) ast.Value {
+	obj := ast.NewObject()
+	for i := 0; i < l1Keys; i++ {
+		l2Obj := ast.NewObject()
+		for j := 0; j < l2Keys; j++ {
+			l3Obj := ast.NewObject()
+			for k := 0; k < l3Keys; k++ {
+				l3Obj.Insert(ast.StringTerm(fmt.Sprintf("%d", k)), ast.BooleanTerm(true))
+			}
+			l2Obj.Insert(ast.StringTerm(fmt.Sprintf("%d", j)), ast.NewTerm(l3Obj))
+		}
+		obj.Insert(ast.StringTerm(fmt.Sprintf("%d", i)), ast.NewTerm(l2Obj))
+	}
+	return obj
+}
+
+// Generates a list of paths for JSON operations. N keys per level, M levels. P patches.
+// TODO: Generate non-conflicting paths.
```

Comment:
what are the side effects of conflicts?

Reply:
Conflicts would potentially cause errors from the EditTree, which are not a problem in normal unit tests, but are a huge problem in benchmarks, because when errors occur, `fmt.Errorf()` rapidly comes to dominate the runtime of the benchmark. :sweat_smile: String formatting can be expensive!